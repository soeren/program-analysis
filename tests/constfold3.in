// Check for constant folding
// - Check inter-procedural folding

program

int x;
int y;
int z;

int identity(int b) { // if called with const arg should return arg
    return b;
}

int givethree(int a) {
    int t;
    int r;
    t := 3;
    r := a*t; // a and t const -> fold to 3
    return t; // t const -> fold to 3 -> function returns const value
}

int diff(int a, int b){ // only const if both arguments are const
    int s;
    int y;
    y := a - b;
    s := y;
    return s;
}

x := diff(4,2); // should be folded to 2, since both args are const
y := 3;

if(y > x){ // both const -> should be folded to true
    x := x*identity(y); // both const -> should be folded to 6
} else {
    x := 0;
}

while(x > 0){ // x not const anymore -> stays the same
    y := y * givethree(0); // y not const anymore, but func -> func is folded
}

x := y - x; // no consts -> stays the same

write(x); // no consts -> stays the same

end
