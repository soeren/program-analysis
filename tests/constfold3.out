program
int x;
int y;
int z;
int identity(int b) // 10
{
     return b; // 11
 }
int givethree(int a) // 12
{
     int t;
     int r;
     t := 3; // 13
     r := a * 3; // 14
     return 3; // 15
 }
int diff(int a,int b) // 16
{
     int s;
     int y;
     y := a - b; // 17
     s := y; // 18
     return s; // 19
 }
x := 2; // 1
y := 3; // 2
if(true) // 3
{
     x := 6; // 4
 } else
{
     x := 0; // 5
 }
while(x > 0) // 6
{
     y := y * 3; // 7
 }
x := y - x; // 8
write(x); // 9
end
